addons/account/models/account_account.py

```diff
@@ -964,7 +964,6 @@ def _adapt_parent_account_group(self, company=None):
                    AND parent.id != child.id
                    AND parent.company_id = child.company_id
                  WHERE child.company_id IN %s
-                   AND child.parent_id IS DISTINCT FROM parent.id -- IMPORTANT avoid to update if nothing changed
```

Comment:
I believe that this line of code was important (as stated by the comment) even though it's quite difficult to measure. So I suggest that we keep the same logic and fix the improvement rather than removing it. We should therefore not update account groups which already have the parent_id set. One way I've possibly found is the following, what do you think ? 

```sql        
WITH relation AS (
    SELECT DISTINCT ON (child.id)
           child.id AS child_id,
           parent.id AS parent_id,
           child.parent_id AS current_parent_id
      FROM account_group parent
      JOIN account_group child
        ON char_length(parent.code_prefix_start) < char_length(child.code_prefix_start)
       AND parent.code_prefix_start <= LEFT(child.code_prefix_start, char_length(parent.code_prefix_start))
       AND parent.code_prefix_end >= LEFT(child.code_prefix_end, char_length(parent.code_prefix_end))
       AND parent.id != child.id
       AND parent.company_id = child.company_id
     WHERE child.company_id IN %s
     ORDER BY child.id, char_length(parent.code_prefix_start) DESC
),
filtered_relation AS (
    -- Filter out any rows where the child already has the correct parent set
    SELECT *
    FROM relation
    WHERE parent_id IS DISTINCT FROM current_parent_id
)
UPDATE account_group child
   SET parent_id = filtered_relation.parent_id
  FROM filtered_relation
 WHERE child.id = filtered_relation.child_id
RETURNING child.id;
```

Reply:
Thank you very much, you made me realize that I did not understand really how the query was working. 
Now, I think there is a much easier way to filter those lines : we can move the check in the `UPDATE`, like this:
```sql
WITH relation AS (
        SELECT DISTINCT ON (child.id)
               child.id AS child_id,
               parent.id AS parent_id
          FROM account_group parent
          JOIN account_group child
            ON char_length(parent.code_prefix_start) < char_length(child.code_prefix_start)
           AND parent.code_prefix_start <= LEFT(child.code_prefix_start, char_length(parent.code_prefix_start))
           AND parent.code_prefix_end >= LEFT(child.code_prefix_end, char_length(parent.code_prefix_end))
           AND parent.id != child.id
           AND parent.company_id = child.company_id
         WHERE child.company_id IN %s
      ORDER BY child.id, char_length(parent.code_prefix_start) DESC
    )
    UPDATE account_group child
       SET parent_id = relation.parent_id
      FROM relation
     WHERE child.id = relation.child_id
       AND child.parent_id IS DISTINCT FROM relation.parent_id
 RETURNING child.id
```
This is what I force-pushed, it passes the test I wrote.
Tell me what you think of it.

Reply:
Indeed it looks even cleaner :+1: